### PR TITLE
perf: various optimizations to eliminate branch misprediction in hash_utils

### PR DESCRIPTION
**Pull request by notashes**
## Which issue does this PR close?

- Part of #20152 

## Rationale for this change

Compile time monomorphization helps bring `rehash` outside the hot loop where it's not required. 

## What changes are included in this PR?

Currently the PR adds a specialized `hash_dictionary_inner()` function with const generic parameters that check for nulls in keys, values. It also handles specific edge cases of just nulls in keys or values.

## Are these changes tested?

There are no additional tests yet. But I will add 'em as I continue. The benchmark results seem promising.
here's `cargo bench --bench with_hashes -- dictionary` for

 <details>
 <summary>origin/main</summary>

```
Gnuplot not found, using plotters backend
Benchmarking dictionary_utf8_int32: single, no nulls
Benchmarking dictionary_utf8_int32: single, no nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: single, no nulls: Collecting 100 samples in estimated 5.0461 s (470k iterations)
Benchmarking dictionary_utf8_int32: single, no nulls: Analyzing
dictionary_utf8_int32: single, no nulls
                        time:   [10.668 µs 10.700 µs 10.734 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

Benchmarking dictionary_utf8_int32: single, nulls
Benchmarking dictionary_utf8_int32: single, nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: single, nulls: Collecting 100 samples in estimated 5.0428 s (409k iterations)
Benchmarking dictionary_utf8_int32: single, nulls: Analyzing
dictionary_utf8_int32: single, nulls
                        time:   [12.269 µs 12.293 µs 12.322 µs]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking dictionary_utf8_int32: multiple, no nulls
Benchmarking dictionary_utf8_int32: multiple, no nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: multiple, no nulls: Collecting 100 samples in estimated 5.0864 s (162k iterations)
Benchmarking dictionary_utf8_int32: multiple, no nulls: Analyzing
dictionary_utf8_int32: multiple, no nulls
                        time:   [31.357 µs 31.426 µs 31.506 µs]
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  1 (1.00%) high severe

Benchmarking dictionary_utf8_int32: multiple, nulls
Benchmarking dictionary_utf8_int32: multiple, nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: multiple, nulls: Collecting 100 samples in estimated 5.0842 s (141k iterations)
Benchmarking dictionary_utf8_int32: multiple, nulls: Analyzing
dictionary_utf8_int32: multiple, nulls
                        time:   [36.060 µs 36.135 µs 36.220 µs]
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe
```
</details>

<details>
<summary>feat/brunch-prediction</summary>

```
Gnuplot not found, using plotters backend
Benchmarking dictionary_utf8_int32: single, no nulls
Benchmarking dictionary_utf8_int32: single, no nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: single, no nulls: Collecting 100 samples in estimated 5.0176 s (1.1M iterations)
Benchmarking dictionary_utf8_int32: single, no nulls: Analyzing
dictionary_utf8_int32: single, no nulls
                        time:   [4.7186 µs 4.7496 µs 4.7821 µs]
                        change: [−55.829% −55.537% −55.240%] (p = 0.00 < 0.05)
                        Performance has improved.

Benchmarking dictionary_utf8_int32: single, nulls
Benchmarking dictionary_utf8_int32: single, nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: single, nulls: Collecting 100 samples in estimated 5.0295 s (712k iterations)
Benchmarking dictionary_utf8_int32: single, nulls: Analyzing
dictionary_utf8_int32: single, nulls
                        time:   [6.9647 µs 7.0426 µs 7.1281 µs]
                        change: [−43.806% −43.445% −42.993%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 16 outliers among 100 measurements (16.00%)
  1 (1.00%) low severe
  4 (4.00%) low mild
  1 (1.00%) high mild
  10 (10.00%) high severe

Benchmarking dictionary_utf8_int32: multiple, no nulls
Benchmarking dictionary_utf8_int32: multiple, no nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: multiple, no nulls: Collecting 100 samples in estimated 5.0600 s (348k iterations)
Benchmarking dictionary_utf8_int32: multiple, no nulls: Analyzing
dictionary_utf8_int32: multiple, no nulls
                        time:   [13.365 µs 13.384 µs 13.404 µs]
                        change: [−57.610% −57.464% −57.313%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) low severe
  4 (4.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe

Benchmarking dictionary_utf8_int32: multiple, nulls
Benchmarking dictionary_utf8_int32: multiple, nulls: Warming up for 3.0000 s
Benchmarking dictionary_utf8_int32: multiple, nulls: Collecting 100 samples in estimated 5.0569 s (242k iterations)
Benchmarking dictionary_utf8_int32: multiple, nulls: Analyzing
dictionary_utf8_int32: multiple, nulls
                        time:   [20.785 µs 20.962 µs 21.173 µs]
                        change: [−42.370% −42.001% −41.579%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  1 (1.00%) low severe
  3 (3.00%) high mild
  14 (14.00%) high severe
```

 </details>

### CODE REVIEW

**Comment by adriangb** [comment]
it would be great to first merge a benchmark into `main` (if one doesn't already exist) to show an improvement

**Comment by notashes** [comment]
hey @adriangb, the benchmarks to test changes already exist in `main`. It was merged with the PR #19373 
The benchmark (`with_hashes`) already measure two key scenarios: 
- single vs multiple columns (for rehash/combine logic)
- with nulls vs without nulls (in keys)

here are some numbers:
```
group                                                main                                   optimized
-----                                                ----                                   ---------
dictionary_utf8_int32: multiple, no nulls            2.73     31.1±0.46µs        ? ?/sec    1.00     11.4±0.83µs        ? ?/sec
dictionary_utf8_int32: multiple, nulls               1.92     39.4±0.61µs        ? ?/sec    1.00     20.5±0.56µs        ? ?/sec
dictionary_utf8_int32: single, no nulls              2.48     10.5±0.19µs        ? ?/sec    1.00      4.2±0.57µs        ? ?/sec
dictionary_utf8_int32: single, nulls                 1.74     12.1±0.18µs        ? ?/sec    1.00      7.0±0.31µs        ? ?/sec
```

But I'm curious do we also want to benchmark cases where the values contain nulls? let me know what you think

**Comment by Dandandan** [comment]
run benchmark with_hashes

**Comment by notashes** [comment]
Hey @adriangb / @Dandandan ,

I have reaised a PR #20182 that adds a couple more benchmark tests that check performance for `StructArray` and `RunArray` related tests. 

and they look like:
```
group                                  main                                    brunch-prediction
-----                                  -----                                 ---------------------
run_array_int32: multiple, no nulls    1.02      5.2±0.08µs        ? ?/sec     1.00      5.1±0.07µs        ? ?/sec
run_array_int32: multiple, nulls       1.03      5.7±0.08µs        ? ?/sec     1.00      5.6±0.05µs        ? ?/sec
run_array_int32: single, no nulls      1.01  1999.2±216.22ns        ? ?/sec    1.00  1986.2±167.83ns        ? ?/sec
run_array_int32: single, nulls         1.03      2.2±0.21µs        ? ?/sec     1.00      2.1±0.16µs        ? ?/sec
struct_array: multiple, no nulls       1.09    117.1±1.10µs        ? ?/sec     1.00    107.6±0.91µs        ? ?/sec
struct_array: multiple, nulls          1.11    141.5±1.68µs        ? ?/sec     1.00    127.5±1.21µs        ? ?/sec
struct_array: single, no nulls         1.08     39.5±0.39µs        ? ?/sec     1.00     36.4±0.44µs        ? ?/sec
struct_array: single, nulls            1.10     47.5±0.44µs        ? ?/sec     1.00     43.1±0.33µs        ? ?/sec
```

**Comment by notashes** [comment]
Hey @Jefffrey,  this is ready for review! Would appreciate your feedback whenever you get time! Thanks

benchmark results (locally tested)

<details>

```
group                                        this-branch                             main
-----                                        ----------                             ---------
dictionary_utf8_int32: multiple, no nulls    1.00     13.9±0.20µs        ? ?/sec    2.33     32.4±0.41µs        ? ?/sec
dictionary_utf8_int32: multiple, nulls       1.00     21.2±0.69µs        ? ?/sec    1.73     36.7±0.43µs        ? ?/sec
dictionary_utf8_int32: single, no nulls      1.00      5.1±0.17µs        ? ?/sec    2.24     11.5±0.28µs        ? ?/sec
dictionary_utf8_int32: single, nulls         1.00      7.3±0.33µs        ? ?/sec    1.72     12.6±0.12µs        ? ?/sec
int64: multiple, no nulls                    1.00     14.8±0.21µs        ? ?/sec    1.02     15.0±0.19µs        ? ?/sec
int64: multiple, nulls                       1.00     25.6±0.30µs        ? ?/sec    1.00     25.6±0.24µs        ? ?/sec
int64: single, no nulls                      1.01      4.6±0.17µs        ? ?/sec    1.00      4.5±0.16µs        ? ?/sec
int64: single, nulls                         1.00      8.2±0.10µs        ? ?/sec    1.01      8.3±0.13µs        ? ?/sec
large_utf8: multiple, no nulls               1.00     63.2±1.04µs        ? ?/sec    1.00     63.1±0.69µs        ? ?/sec
large_utf8: multiple, nulls                  1.01     71.0±1.16µs        ? ?/sec    1.00     70.3±0.66µs        ? ?/sec
large_utf8: single, no nulls                 1.00     20.5±0.24µs        ? ?/sec    1.01     20.8±0.18µs        ? ?/sec
large_utf8: single, nulls                    1.00     23.0±0.28µs        ? ?/sec    1.00     22.9±0.71µs        ? ?/sec
run_array_int32: multiple, no nulls          1.00      4.0±0.09µs        ? ?/sec    1.34      5.3±0.05µs        ? ?/sec
run_array_int32: multiple, nulls             1.00      4.9±0.08µs        ? ?/sec    1.13      5.5±0.09µs        ? ?/sec
run_array_int32: single, no nulls            1.00      2.0±0.26µs        ? ?/sec    1.18      2.4±0.18µs        ? ?/sec
run_array_int32: single, nulls               1.00      2.4±0.25µs        ? ?/sec    1.01      2.4±0.19µs        ? ?/sec
struct_array: multiple, no nulls             1.00    124.7±0.98µs        ? ?/sec    1.09    136.3±1.16µs        ? ?/sec
struct_array: multiple, nulls                1.00    146.9±1.44µs        ? ?/sec    1.10    161.3±1.75µs        ? ?/sec
struct_array: single, no nulls               1.00     42.3±0.60µs        ? ?/sec    1.09     45.9±0.64µs        ? ?/sec
struct_array: single, nulls                  1.00     49.8±0.50µs        ? ?/sec    1.09     54.4±0.60µs        ? ?/sec
utf8: multiple, no nulls                     1.00     61.6±0.60µs        ? ?/sec    1.01     62.2±0.92µs        ? ?/sec
utf8: multiple, nulls                        1.02     71.1±1.55µs        ? ?/sec    1.00     70.0±0.69µs        ? ?/sec
utf8: single, no nulls                       1.00     20.7±0.30µs        ? ?/sec    1.00     20.7±0.21µs        ? ?/sec
utf8: single, nulls                          1.01     22.9±1.25µs        ? ?/sec    1.00     22.6±0.26µs        ? ?/sec
utf8_view (small): multiple, no nulls        1.00     19.5±0.24µs        ? ?/sec    1.00     19.6±0.22µs        ? ?/sec
utf8_view (small): multiple, nulls           1.00     29.2±0.18µs        ? ?/sec    1.01     29.5±0.28µs        ? ?/sec
utf8_view (small): single, no nulls          1.02      5.5±0.08µs        ? ?/sec    1.00      5.3±0.08µs        ? ?/sec
utf8_view (small): single, nulls             1.00      9.1±0.07µs        ? ?/sec    1.01      9.2±0.08µs        ? ?/sec
utf8_view: multiple, no nulls                1.00     54.9±0.73µs        ? ?/sec    1.02     56.2±1.12µs        ? ?/sec
utf8_view: multiple, nulls                   1.02     63.0±0.77µs        ? ?/sec    1.00     61.9±0.72µs        ? ?/sec
utf8_view: single, no nulls                  1.03     18.1±0.40µs        ? ?/sec    1.00     17.7±0.28µs        ? ?/sec
utf8_view: single, nulls                     1.01     19.9±0.21µs        ? ?/sec    1.00     19.6±0.37µs        ? ?/sec
```

</details>

**Comment by adriangb** [comment]
@notashes can you rebase now that https://github.com/apache/datafusion/pull/20182 is merged?

**Comment by notashes** [comment]
@adriangb done! thank you!

**Comment by adriangb** [comment]
run benchmark with_hashes

**Comment by adriangb** [comment]
I’ll leave the review up to @Jefffrey but from my perspective benchmarks look good and cursory glance at change looks well structured. This is worth reviewing and moving forward.

File: datafusion/common/src/hash_utils.rs [hunk 1]
```diff
@@ -427,22 +416,93 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
...
+                    } else {
+                        *hash = dict_hashes[idx];
+                    }
+                }
+            }
+        }
+    } else {
+        for (hash, key) in hashes_buffer.iter_mut().zip(array.keys().iter()) {
+            // SAFETY: Keys are guaranteed to be non-null because we checked nucll_count = 0
+            let key = key.unwrap();
```
Comment (Jefffrey) [inline datafusion/common/src/hash_utils.rs]:
```suggestion
        for (hash, key) in hashes_buffer.iter_mut().zip(array.keys().values()) {
```

Can directly go to the values buffer here

**Review by Jefffrey — APPROVED** [review]
These changes look good with some nice benchmark improvements 👍 

I've changed the PR body to not close the original issue however, as there are still some points from it to address in follow ups
